File: python/exercises/practice/unit_test_3632ff20_3eea_4004_9156_3ebbdca8ab1c/event_scheduler/.meta/event_scheduler.py
import threading
import heapq

_lock = threading.Lock()
_event_map = {}
_event_heap = []

def schedule_event(event_id, execution_time, payload, priority):
    global _event_map, _event_heap
    with _lock:
        if event_id in _event_map:
            return False
        _event_map[event_id] = (execution_time, payload, priority)
        heapq.heappush(_event_heap, (execution_time, event_id))
    return True

def cancel_event(event_id):
    global _event_map
    with _lock:
        if event_id in _event_map:
            del _event_map[event_id]
            return True
    return False

def get_next_events(current_time, max_events):
    global _event_map, _event_heap
    eligible_candidates = []
    remaining_candidates = []
    with _lock:
        # Extract all eligible events from the heap
        while _event_heap and _event_heap[0][0] <= current_time:
            exec_time, event_id = heapq.heappop(_event_heap)
            if event_id not in _event_map:
                continue
            # Retrieve the current event details
            et, payload, priority = _event_map[event_id]
            eligible_candidates.append((priority, exec_time, event_id, payload))
        
        # Sort the eligible events:
        # First by descending priority and then by ascending execution time.
        eligible_candidates.sort(key=lambda x: (-x[0], x[1]))
        
        # Select up to max_events events to return and reinsert the remainder
        selected_candidates = eligible_candidates[:max_events]
        remaining_candidates = eligible_candidates[max_events:]
        
        for candidate in selected_candidates:
            _, _, event_id, _ = candidate
            if event_id in _event_map:
                del _event_map[event_id]
        
        for candidate in remaining_candidates:
            _, exec_time, event_id, _ = candidate
            heapq.heappush(_event_heap, (exec_time, event_id))
    
    result = [(event_id, payload) for (_, _, event_id, payload) in selected_candidates]
    return result

File: go/exercises/practice/unit_test_06e5ca8d_3b0f_4748_92b2_f8f2a5a74672/event_scheduler/.meta/event_scheduler.go
package event_scheduler

import (
	"container/heap"
	"errors"
	"fmt"
	"sync"
	"time"
)

// Event represents a scheduled event with metadata
type Event struct {
	ID            string
	ExecutionTime int64
	Payload       string
	Executed      bool
	Cancelled     bool
}

// EventScheduler is the main scheduler struct
type EventScheduler struct {
	events      map[string]*Event
	timeIndex   *TimeHeap
	mutex       sync.RWMutex
}

// NewEventScheduler creates a new instance of EventScheduler
func NewEventScheduler() *EventScheduler {
	scheduler := &EventScheduler{
		events:    make(map[string]*Event),
		timeIndex: &TimeHeap{},
	}
	heap.Init(scheduler.timeIndex)
	return scheduler
}

// ScheduleEvent schedules a new event or updates an existing one
func (es *EventScheduler) ScheduleEvent(eventID string, executionTime int64, payload string) error {
	if eventID == "" {
		return errors.New("event ID cannot be empty")
	}

	// Check if execution time is in the past
	if executionTime <= time.Now().Unix() {
		return errors.New("execution time must be in the future")
	}

	es.mutex.Lock()
	defer es.mutex.Unlock()

	// Check if event already exists (idempotence)
	existingEvent, exists := es.events[eventID]
	if exists {
		// Update the existing event
		if !existingEvent.Executed && !existingEvent.Cancelled {
			existingEvent.ExecutionTime = executionTime
			existingEvent.Payload = payload
			
			// Re-heapify to account for the updated execution time
			heap.Init(es.timeIndex)
		}
		return nil
	}

	// Create a new event
	newEvent := &Event{
		ID:            eventID,
		ExecutionTime: executionTime,
		Payload:       payload,
		Executed:      false,
		Cancelled:     false,
	}

	// Add to map and priority queue
	es.events[eventID] = newEvent
	heap.Push(es.timeIndex, eventIDTime{
		ID:   eventID,
		Time: executionTime,
	})

	return nil
}

// CancelEvent cancels a previously scheduled event
func (es *EventScheduler) CancelEvent(eventID string) error {
	if eventID == "" {
		return errors.New("event ID cannot be empty")
	}

	es.mutex.Lock()
	defer es.mutex.Unlock()

	event, exists := es.events[eventID]
	if !exists {
		return fmt.Errorf("event with ID %s does not exist", eventID)
	}

	if event.Executed {
		return fmt.Errorf("event with ID %s has already been executed", eventID)
	}

	event.Cancelled = true
	return nil
}

// GetNextEvents retrieves events scheduled at or before the given time
func (es *EventScheduler) GetNextEvents(currentTime int64, limit int) ([]string, error) {
	if limit < 0 {
		return nil, errors.New("limit must be non-negative")
	}

	if limit == 0 {
		return []string{}, nil
	}

	es.mutex.RLock()
	defer es.mutex.RUnlock()

	// Create a temporary copy of the time index for querying
	tempHeap := &TimeHeap{}
	*tempHeap = append(*tempHeap, (*es.timeIndex)...)
	heap.Init(tempHeap)

	result := make([]string, 0, limit)
	for tempHeap.Len() > 0 && len(result) < limit {
		item := heap.Pop(tempHeap).(eventIDTime)
		
		// Check if event is scheduled at or before currentTime
		if item.Time > currentTime {
			break
		}

		// Retrieve event details
		event, exists := es.events[item.ID]
		if exists && !event.Executed && !event.Cancelled {
			result = append(result, item.ID)
		}
	}

	return result, nil
}

// MarkEventAsExecuted marks an event as executed
func (es *EventScheduler) MarkEventAsExecuted(eventID string) error {
	if eventID == "" {
		return errors.New("event ID cannot be empty")
	}

	es.mutex.Lock()
	defer es.mutex.Unlock()

	event, exists := es.events[eventID]
	if !exists {
		return fmt.Errorf("event with ID %s does not exist", eventID)
	}

	// Idempotence: if already executed, just return success
	if event.Executed {
		return nil
	}

	if event.Cancelled {
		return fmt.Errorf("event with ID %s has been cancelled", eventID)
	}

	event.Executed = true
	return nil
}

// TimeHeap is a min-heap of event IDs ordered by execution time
type TimeHeap []eventIDTime

type eventIDTime struct {
	ID   string
	Time int64
}

// Implementation of heap.Interface for TimeHeap
func (h TimeHeap) Len() int { return len(h) }

func (h TimeHeap) Less(i, j int) bool {
	return h[i].Time < h[j].Time
}

func (h TimeHeap) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
}

func (h *TimeHeap) Push(x interface{}) {
	*h = append(*h, x.(eventIDTime))
}

func (h *TimeHeap) Pop() interface{} {
	old := *h
	n := len(old)
	item := old[n-1]
	*h = old[0 : n-1]
	return item
}